clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java

```diff
@@ -34,4 +43,25 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
         return new KafkaPrincipal(split[0], split[1]);
     }
 
+    public static void addConfiguredSecurityProviders(Map<String, ?> configs) {
+        String securityProviderClassesStr = (String) configs.get(SecurityConfig.SECURITY_PROVIDERS_CONFIG);
+        if (securityProviderClassesStr == null || securityProviderClassesStr.equals("")) {
+            return;
+        }
+        try {
+            String[] securityProviderClasses = securityProviderClassesStr.replaceAll("\\s+", "").split(",");
+            for (int index = 0; index < securityProviderClasses.length; index++) {
+                SecurityProviderGenerator securityProviderGenerator = (SecurityProviderGenerator) Class.forName(securityProviderClasses[index]).newInstance();
+                Security.insertProviderAt(securityProviderGenerator.getProvider(), index + 1);
+            }
+        } catch (ClassCastException e) {
+            LOGGER.error("Generator provided through " + SecurityConfig.SECURITY_PROVIDERS_CONFIG +
+                    " are expected to be sub-classes of SecurityProviderGenerator");
+        } catch (ClassNotFoundException cnfe) {
+            LOGGER.error("Unrecognized security provider class", cnfe);
+        } catch (IllegalAccessException | InstantiationException e) {
+            LOGGER.error("Unexpected implementation of security provider class", e);
+        }
+    }
+
```

Comment:
Suggestion:
1.
I think using following format for security.providers:
security.providers=provider1_classname,provider2_generator_classname key1:val1 key2:val2,...
2.
So when parsing above config, if there is no parameters following provider1_classname, then we can think provider1_classname is java Provider, then insert it; if there are key:value pair parameters following provider2_generator_classname, we can think provider2_generator_classname is SecurityProviderGenerator, then create "Map<String, ?> config" from key:value pair parameters, then call configure and getProvider of instance of SecurityProviderGenerator. This way we can handle all different scenarios in the future.

Reply:
@jeffchao traditionally Kafka used key,value pairs in properties and pass it everywhere and each implementation takes look at this config and pulls their interested key,value pairs.
Example, authorizer interface https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/security/auth/Authorizer.scala#L35 . The pluggable class when it gets instantiated a configure method will be called and all the key,value in server.properties will be passed and it will pick whats relevant to the class.
We can do the same here instead of asking users append key,values into the a config which is hard to configure and hard to get it right.

Reply:
I got what you said. But for this case, key/value pairs might be very specific to the provider. 
1. Do you mean we can configure SecurityProviderGenerator like JAAS that will configure class name and key/value pairs parameters together? 
2. How do we configure a Provider which will not need key/value pairs?  Do we require a wrapper which will  implement SecurityProviderGenerator for the Provider?

Reply:
1. No, we're not going with that. You can configure any properties you need in the server.properties file as key value pairs just like other config params(not like JAAS) and all the config param will be passed to the configure method of the SecurityProviderGenerator(renamed to SecurityProviderCreator). Your implementation of SecurityProviderCreator can read only the config that makes sense to it.
2. All the providerCreator classes that don't need the configure method can skip overriding this method. A default implementation of configure method is added in SecurityProviderCreator.

Reply:
ok. Just curious what is target date when all implementation for this will be merged?

Reply:
It really depends on when the PR gets accepted. I'll try to merge this as soon as it gets accepted.

Reply:
It seems I also could approve it. I will read all code tomorrow and work with you to get this approved.

Reply:
I do not see default implementation of configure method is added in SecurityProviderCreator?